### PR TITLE
Buffered reading and writing MRD files, resulting in 6x speedup on read and 4x on write

### PR DESCRIPTION
**Pull request by dchansen**
This PR makes HDF5 reuse the internal buffers used for reading from a dataset.
This reduces the time it takes to read individual acquisitions by 6x (at least on Windows) and makes reading individual acquisitions as fast as reading the entire dataset in one go.

The cost is that datasets now carry 2Kb of buffer state, which I think is acceptable.

### CODE REVIEW

**Comment by dchansen** [comment]
On linux, the improvement is more modest - on the order of 30%.

File: tests/test_dataset.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+float create_random_float() {
+    return dist(rng);
+}
+
+BOOST_AUTO_TEST_SUITE(DatasetTest)
+
+
+BOOST_AUTO_TEST_CASE(test_acquisition_header) {
+
```
Comment (hansenms) [inline tests/test_dataset.cpp:22]: nit: formatting and additional line breaks in this file.

**Review by hansenms — CHANGES_REQUESTED** [review]
Look OK in general. I mainly have some questions on whether this should be optional and what the right size of that buffer is. Maybe if those things were configurable, it would be easier to test with the benchmark app? Also some thoughts on handling versioning with the adde vcpkg configuration.

File: vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{
+  "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/vcpkg.schema.json",
+  "name": "ismrmrd",
+  "version": "1.12.0",
```
[the remark this answers is not in the document]
Comment (dchansen) [reply vcpkg.json]: Managed to get CMake to actually check that this matches.

File: tests/benchmark_dataset.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+static boost::random::mt19937 rng;
+static boost::random::uniform_real_distribution<float> dist = boost::random::uniform_real_distribution<float>();
+
+float create_random_float() {
+    return dist(rng);
+}
+
+int main(int argc, char** argv) {
+
```
[the remark this answers is not in the document]
Comment (dchansen) [reply tests/benchmark_dataset.cpp:18]: Done

**Comment by dchansen** [comment]
@hansenms This should be ready to merge I think, unless you have any other comments?

File: libsrc/dataset.c [hunk 4]
```diff
@@ -152,6 +152,11 @@ static int delete_var(const ISMRMRD_Dataset *dset, const char *var) {
     return status;
 }
 
+#define ISMRMRD_READ_BUFFER_SIZE 1024*1024
+
+
+static char ismrmrd_conversion_buffer[ISMRMRD_READ_BUFFER_SIZE];
```
[the remark this answers is not in the document]
Comment (dchansen) [reply libsrc/dataset.c:158]: Yes. This is baked into HDF5. The "multithreaded" version is just using a global lock.

File: libsrc/dataset.c [hunk 5]
```diff
@@ -152,6 +152,11 @@ static int delete_var(const ISMRMRD_Dataset *dset, const char *var) {
     return status;
 }
 
+#define ISMRMRD_READ_BUFFER_SIZE 1024*1024
```
Comment (johnstairs) [inline libsrc/dataset.c:155]: How did you arrive at this number?

File: tests/benchmark_dataset.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    for (size_t i = 0; i < acqs.size(); i++) {
+        std::generate((float *)acqs[i].data_begin(), (float *)acqs[i].data_end(), create_random_float);
+        std::generate((float *)acqs[i].traj_begin(), (float *)acqs[i].traj_end(), create_random_float);
+    }
+
+    {
+
+        auto start = std::chrono::high_resolution_clock::now();
+        Dataset dataset = Dataset(temp.string().c_str(), "/test", true);
+        for (size_t i = 0; i < acqs.size(); i++)
```
Comment (hansenms) [inline tests/benchmark_dataset.cpp:33]: nit: Added `{}` to the for loop.

File: tests/test_dataset.cpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+static boost::random::uniform_real_distribution<float> dist = boost::random::uniform_real_distribution<float>();
+
+float create_random_float() {
+    return dist(rng);
+}
+
+
+BOOST_AUTO_TEST_SUITE(DatasetTest)
+
+BOOST_AUTO_TEST_CASE(test_acquisition_header) {
```
Comment (hansenms) [inline tests/test_dataset.cpp:20]: Think that test name is copied from the acquisition tests. Maybe choose a better name.

**Review by hansenms — APPROVED** [review]
Some minor additional comments, but otherwise looks good.

File: libsrc/dataset.c (hunk 5 above)
Comment (dchansen) [reply libsrc/dataset.c:155]: It's the same as the default value in HDF5. 
I haven't played around with it too much, but 1024 seemed to give some degradation in write performance on linux.

**Pull request merged**